library/core/src/str/count.rs

```diff
@@ -0,0 +1,116 @@
+//! Code for efficiently counting the number of `char`s in a UTF-8 encoded
+//! string.
+//!
+//! Broadly, UTF-8 encodes `char`s as a "leading" byte which begins the `char`,
+//! followed by some number (possibly 0) of continuation bytes.
+//!
+//! The leading byte can have a number of bit-patterns (with the specific
+//! pattern indicating how many continuation bytes follow), but the continuation
+//! bytes are always in the format `0b10XX_XXXX` (where the `X`s can take any
+//! value). That is, the most significant bit is set, and the second most
+//! significant bit is unset.
+//!
+//! To count the number of characters, we can just count the number of bytes in
+//! the string which are not continuation bytes, which can be done many bytes at
+//! a time fairly easily.
+//!
+//! Note: Because the term "leading byte" can sometimes be ambiguous (for
+//! example, it could also refer to the first byte of a slice), we'll often use
+//! the term "non-continuation byte" to refer to these bytes in the code.
+
+pub(super) fn count_chars(s: &str) -> usize {
+    // For correctness, `CHUNK_SIZE` must be:
+    // - Less than or equal to 255, otherwise we'll overflow bytes in `counts`.
+    // - A multiple of `UNROLL_INNER`, otherwise our `break` inside the
+    //   `body.chunks(CHUNK_SIZE)` loop.
+    //
+    // For performance, `CHUNK_SIZE` should be:
+    // - Relatively cheap to `%` against.
+    // - Large enough to avoid paying for the cost of the `sum_bytes_in_usize`
+    //   too often.
+    const CHUNK_SIZE: usize = 192;
+    const UNROLL_INNER: usize = 4;
+
+    // Check the properties of `CHUNK_SIZE` / `UNROLL_INNER` that are required
+    // for correctness.
+    const _: [(); 1] = [(); (CHUNK_SIZE < 256 && (CHUNK_SIZE % UNROLL_INNER) == 0) as usize];
+    // SAFETY: transmuting `[u8]` to `[usize]` is safe except for size
+    // differences which are handled by `align_to`.
+    let (head, body, tail) = unsafe { s.as_bytes().align_to::<usize>() };
+
+    let mut total = char_count_general_case(head) + char_count_general_case(tail);
+    // Split `body` into `CHUNK_SIZE` chunks to reduce the frequency with which
+    // we call `sum_bytes_in_usize`.
+    for chunk in body.chunks(CHUNK_SIZE) {
+        // We accumulate intermediate sums in `counts`, where each byte contains
+        // a subset of the sum of this chunk, like a `[u8; size_of::<usize>()]`.
+        let mut counts = 0;
+        let unrolled_chunks = chunk.array_chunks::<UNROLL_INNER>();
+        // If there's a remainder (know can only happen for the last item in
+        // `chunks`, because `CHUNK_SIZE % UNROLL == 0`), then we need to
+        // account for that (although we don't use it to later).
+        let remainder = unrolled_chunks.remainder();
```

Comment:
Doing something like this may read better:

```
let (unrolled_chunks, remainder) = chunk.as_chunks::<UNROLL_INNER>();
...
```

or perhaps changing line 39 above to `s.as_bytes().align_to::<[usize; UNROLL_INNER]>()` might work well too I think?

Reply:
Whoa, TIL as_chunks, that API is way nicer!